src/riscv/lib/src/pvm/linux/signals.rs

```diff
@@ -16,20 +18,133 @@ use crate::state::NewState;
 use crate::state_backend::AllocatedOf;
 use crate::state_backend::Atom;
 use crate::state_backend::Cell;
+use crate::state_backend::Elem;
 use crate::state_backend::FnManager;
 use crate::state_backend::ManagerAlloc;
 use crate::state_backend::ManagerBase;
 use crate::state_backend::ManagerClone;
+use crate::state_backend::ManagerRead;
 use crate::state_backend::ManagerReadWrite;
 use crate::state_backend::Ref;
 use crate::struct_layout;
 
+#[repr(C)]
+#[derive(Clone, Debug)]
+/// Linux sigaction struct, see <https://man7.org/linux/man-pages/man2/sigaction.2.html>
+pub struct LinuxSigAction {
+    sa_handler: VirtAddr,
+    sa_sigaction: VirtAddr,
+    sa_mask: u32,
+    sa_flags: u32,
+    sa_restorer: VirtAddr,
+}
+
+#[cfg(test)]
+// Currently we only support the `sa_sigaction` field, so in tests it's useful to be able to
+// create a sigaction using only this field.
+impl LinuxSigAction {
+    pub(crate) fn new(sa_sigaction: VirtAddr) -> Self {
+        Self {
+            sa_handler: VirtAddr::new(0),
+            sa_sigaction,
+            sa_mask: 0,
+            sa_flags: 0,
+            sa_restorer: VirtAddr::new(0),
+        }
+    }
+}
+
+#[cfg(test)]
+impl Default for LinuxSigAction {
+    fn default() -> Self {
+        Self {
+            sa_handler: VirtAddr::new(0),
+            sa_sigaction: VirtAddr::new(0),
+            sa_mask: 0,
+            sa_flags: 0,
+            sa_restorer: VirtAddr::new(0),
+        }
+    }
+}
+
+#[cfg(test)]
+impl PartialEq for LinuxSigAction {
+    fn eq(&self, other: &Self) -> bool {
+        self.sa_handler == other.sa_handler
+            && self.sa_sigaction == other.sa_sigaction
+            && self.sa_mask == other.sa_mask
+            && self.sa_flags == other.sa_flags
+            && self.sa_restorer == other.sa_restorer
+    }
+}
+
 /// `size_of(struct sigaction)` on the Kernel side
 const SIZE_SIGACTION: usize = 32;
 
-// For [Cell]<E, _>, `E` must be 'static. For this reason, each field of the linux sigaction
-// struct will have its own array of the primitives or wrappers around primitives (e.g.
-// [VirtAddr]) used for the member's type.
+impl Elem for LinuxSigAction {
+    const STORED_SIZE: NonZeroUsize = NonZeroUsize::new(SIZE_SIGACTION).unwrap();
+
+    unsafe fn read_unaligned(source: *const u8) -> Self {
+        unsafe {
```

Comment:
These functions should have safety comments

Reply:
Sorry, I meant to say that the `unsafe` blocks inside the functions should have safety comments, in this case explaining why these casts are safe.

Reply:
Sorry, I meant to say that the `unsafe` blocks inside the functions should have safety comments, in this case explaining why these casts are safe.